pkg/controller/volume/attachdetach/cache/actual_state_of_world.go

```diff
@@ -209,6 +210,11 @@ type actualStateOfWorld struct {
 	// the node (including the list of volumes to report attached).
 	nodesToUpdateStatusFor map[types.NodeName]nodeToUpdateStatusFor
 
+	// nodesAttachedVolumes is map containing the set for volumes for each node and
+	// in sync with attachedVolumes. The key in this map is the name of the node
+	// and the value is all volumes the node have attached.
+	nodesAttachedVolumes map[types.NodeName]map[v1.UniqueVolumeName]AttachedVolume
```

Comment:
My primary concern is with introduction with introduction of this new variable is, it doesn't actually get synced with rest of the attach-detach controller code. For example, when `SetVolumeMountedByNode` is called, `AttachedVolume` inside `nodesAttachedVolumes` will still report old value, because it doesn't actually get updated.

This will create problems in future maintenance of the controller, because now we have this duplicate state which doesn't really get synced and use of this state in other places will be dangerous.

Reply:
Nice observation and  it is my concern also！Saparated cache state will cause out of sync problem. I will try to update the code to make sure they are in sync for each modification and  add comment to hint future developer.